src/clusterfuzz/_internal/google_cloud_utils/batch.py

```diff
@@ -71,48 +77,87 @@ def get_job_name():
+    job_specs[spec].append(batch_task.input_download_url)
+
+  return [
+      _create_job(spec, input_urls) for spec, input_urls in job_specs.items()
+  ]
+
+
+def _get_task_spec(batch_workload_spec):
+  """Gets the task spec based on the batch workload spec."""
   runnable = batch.Runnable()
   runnable.container = batch.Runnable.Container()
-  spec = get_spec(module_name, cf_job)
-  runnable.container.image_uri = spec.docker_image
+  runnable.container.image_uri = batch_workload_spec.docker_image
   runnable.container.options = (
-      '--memory-swappiness=40 --shm-size=1.9g --rm --net=host -e HOST_UID=1337 '
-      '-P --privileged --cap-add=all '
+      '--memory-swappiness=40 --shm-size=1.9g --rm --net=host '
+      '-e HOST_UID=1337 -P --privileged --cap-add=all '
       '--name=clusterfuzz -e UNTRUSTED_WORKER=False -e UWORKER=True '
-      f'-e UWORKER_INPUT_DOWNLOAD_URL={input_download_url}')
+      '-e UWORKER_INPUT_DOWNLOAD_URL')
   runnable.container.volumes = ['/var/scratch0:/mnt/scratch0']
-  # Jobs can be divided into tasks. In this case, we have only one task.
-  task = batch.TaskSpec()
-  task.runnables = [runnable]
-  task.max_retry_count = RETRY_COUNT
+  task_spec = batch.TaskSpec()
+  task_spec.runnables = [runnable]
+  task_spec.max_retry_count = RETRY_COUNT
   # TODO(metzman): Change this for production.
-  task.max_run_duration = MAX_DURATION
+  task_spec.max_run_duration = MAX_DURATION
+  return task_spec
 
-  # Only one of these is currently possible.
-  group = batch.TaskGroup()
-  group.task_count = TASK_COUNT
-  group.task_spec = task
 
-  policy = batch.AllocationPolicy.InstancePolicy()
+def _get_allocation_policy(spec):
+  """Returns the allocation policy for a BatchWorkloadSpec."""
+  instance_policy = batch.AllocationPolicy.InstancePolicy()
   disk = batch.AllocationPolicy.Disk()
   disk.image = 'batch-cos'
   disk.size_gb = spec.disk_size_gb
   disk.type = spec.disk_type
-  policy.boot_disk = disk
-  policy.machine_type = spec.machine_type
+  instance_policy.boot_disk = disk
+  instance_policy.machine_type = spec.machine_type
   instances = batch.AllocationPolicy.InstancePolicyOrTemplate()
-  instances.policy = policy
+  instances.policy = instance_policy
   allocation_policy = batch.AllocationPolicy()
   allocation_policy.instances = [instances]
   service_account = batch.ServiceAccount(email=spec.service_account_email)  # pylint: disable=no-member
   allocation_policy.service_account = service_account
+  return allocation_policy
+
+
+def _create_job(spec, input_urls):
+  """Creates and starts a batch job from |spec| that executes all tasks."""
+  task_group = batch.TaskGroup()
+  task_group.task_count = len(input_urls)
+  assert task_group.task_count < MAX_CONCURRENT_VMS_PER_JOB
+  task_environments = [
+      batch.Environment(variables={'UWORKER_INPUT_DOWNLOAD_URL': input_url})
+      for input_url in input_urls
+  ]
+  task_group.task_environments = task_environments
```

Comment:
This may not end up mattering much, but I assume the individual tasks can't steal input URLs from other tasks?

Reply:
Yes they can't steal them.

Reply:
Although, the nice thing here is that the tasks are all based on the same testcase. So there's nothing for any of them to steal. Defense in depth!